lib/pengine/rules.c

```diff
@@ -1160,44 +1182,40 @@ pe__eval_op_expr(const xmlNode *expr, const pe_rule_eval_data_t *rule_data)
 gboolean
 pe__eval_role_expr(const xmlNode *expr, const pe_rule_eval_data_t *rule_data)
 {
-    gboolean accept = FALSE;
     const char *op = NULL;
-    const char *value = NULL;
 
-    if (rule_data->role == pcmk_role_unknown) {
-        return accept;
+    if (rule_data->role == pcmk_role_unknown) { // Skip role checks for now
+        return FALSE;
     }
 
-    value = crm_element_value(expr, XML_EXPR_ATTR_VALUE);
     op = crm_element_value(expr, XML_EXPR_ATTR_OPERATION);
 
+    // @TODO Why are only Unpromoted/Promoted considered defined?
+
     if (pcmk__str_eq(op, "defined", pcmk__str_casei)) {
         if (rule_data->role > pcmk_role_started) {
-            accept = TRUE;
+            return TRUE;
         }
 
     } else if (pcmk__str_eq(op, "not_defined", pcmk__str_casei)) {
         if ((rule_data->role > pcmk_role_unknown)
             && (rule_data->role < pcmk_role_unpromoted)) {
-            accept = TRUE;
+            return TRUE;
         }
 
     } else if (pcmk__str_eq(op, "eq", pcmk__str_casei)) {
-        if (text2role(value) == rule_data->role) {
-            accept = TRUE;
-        }
+        return role_matches(expr, rule_data)? TRUE : FALSE;
 
-    } else if (pcmk__str_eq(op, "ne", pcmk__str_casei)) {
-        // Test "ne" only with promotable clone roles
-        if ((rule_data->role > pcmk_role_unknown)
-            && (rule_data->role < pcmk_role_unpromoted)) {
-            accept = FALSE;
+    } else if (pcmk__str_eq(op, "ne", pcmk__str_casei)
+               // Test "ne" only with promotable clone roles (@TODO Why?)
+               && (rule_data->role >= pcmk_role_unpromoted)) {
+        return role_matches(expr, rule_data)? FALSE : TRUE;
```

Comment:
It's moot since the code is dead, going to ignore (though I suspect the new behavior would be an improvement in error-checking if #role were actually implemented)